File: frontend/ios/Hive/Hive/StatusOr.swift
//
//  StatusOr.swift
//  Hive
//
//  Created by Chuck Onwuzuruike on 10/4/18.
//  Copyright © 2018 Chuck Onwuzuruike. All rights reserved.
//

import Foundation

class StatusOr<T> {
    
    private(set) var errorMessage: String?
    private(set) var error: StatusError?
    private(set) var value: T?
    
    init(v: T?, err: StatusError?, msg: String?) {
        self.value = v
        self.error = err == nil ? StatusError.NO_ERROR : err
        self.errorMessage = msg == nil ? "" : msg
    }
    
    convenience init(err: StatusError?, msg: String?) {
        self.init(v: Optional.none, err: err, msg: msg)
    }
    
    convenience init(err: StatusError?) {
        self.init(v: Optional.none, err: err, msg: Optional.none)
    }
    
    convenience init(v: T?) {
        self.init(v: v, err: Optional.none, msg: Optional.none)
    }
    
    public func get() -> T {
        if (self.value == nil) {
            fatalError("No value to get from. Error: " + getErrorMessage())
        }
        return self.value!
    }
    
    public func hasError() -> Bool {
        // Error is never nil... but sometimes you just never know.
        return error != nil || error! != StatusError.NO_ERROR
    }
    
    public func getStatusError() -> StatusError {
        return error!
    }
    
    public func getErrorMessage() -> String {
        return String(describing: getStatusError()) + " : " + errorMessage!
    }
    
    public var description: String {
        if (hasError()) {
            return getErrorMessage()
        }
        return value.debugDescription
    }
}

File: clients/java/Hive/src/hive/StatusOr.java
package hive;

public class StatusOr<T> {
	
	private String errorMessage;
	private StatusError error;
	private T value;
	
	protected StatusOr(T v, StatusError err, String msg) {
		errorMessage = msg;
		error = err;
		value = v;
	}
	
	@SuppressWarnings({ "rawtypes", "unchecked" })
	public StatusOr(StatusOr statusOr) {
		if (statusOr.hasError()) {
			errorMessage = statusOr.errorMessage;
			error = statusOr.error;
		} else {
			value = (T) statusOr.get();
		}
	}
	
	public StatusOr(T v) {
		this(v, StatusError.NO_ERROR, "");
	}
	
	public StatusOr(StatusError err, String msg) {
		this(null, err, msg);
	}
	
	public StatusOr(StatusError err) {
		this(null, err, "");
	}
	
	public T get() {
		if (hasError()) {
			RuntimeException e = new RuntimeException("Unhandled error: \n" + getErrorMessage() + "\n");
			e.printStackTrace();
		}
		return value;
	}
	
	public boolean hasError() {
		return !error.equals(StatusError.NO_ERROR);
	}
	
	public StatusError getStatusError() {
		return error;
	}
	
	public String getErrorMessage() {
		return error.name() + ": " + errorMessage;
	}
	
	public String toString() {
		if (hasError()) {
			return getErrorMessage();
		}
		return value.toString();
	}
}
